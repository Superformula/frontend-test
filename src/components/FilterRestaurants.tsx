import * as React from 'react'
import { Fragment } from 'react'

import Dropdown from './Dropdown'
import { Cost, Categories } from '../models/models'

export default function FilterRestaurants() {
  return (
    <Fragment>
      {"Filter By:" }
      {/* Open Now buttom */}
      <Dropdown className="Price" ddList={Cost} ddHeader="Price"></Dropdown>
      <Dropdown className="Categories" ddList={Categories} ddHeader="Categories"></Dropdown>
      {/* Clear All button */}
    </Fragment>
  )
}
