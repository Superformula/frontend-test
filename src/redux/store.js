import { createStore, applyMiddleware } from 'redux'
import logger from 'redux-logger'
import rootReducer from './root.reducer'

const middlewares = __IS_PROD__ ? [] : [logger]

const store = createStore(rootReducer, applyMiddleware(...middlewares))

export default store
