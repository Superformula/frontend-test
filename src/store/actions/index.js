import {
	FETCH_RESTAURANTS_LOADING,
	FETCH_RESTAURANTS_SUCCESS,
	FETCH_REVIEWS_LOADING,
	FETCH_REVIEWS_SUCCESS,
	FETCH_RESTAURANT_DETAILS_LOADING,
	FETCH_RESTAURANT_DETAILS_SUCCESS,
	OPEN_NOW_CHANGED,
	SELECTED_PRICE_CHANGED,
	SELECTED_CATEGORY_CHANGED,
	QUERY_OFFSET_CHANGED,
	CLEAR_FILTERS,
	CLEAR_RESTAURANTS,
	FETCH_ERROR
} from '../constants';

export const OFFSET_INCREMENT = 20;

import { fetchRestaurants, fetchRestaurantDetails, fetchReviews } from '../../services';

export const loadError = error => {
	return {
		type: FETCH_ERROR,
		value: error
	};
};

export const getRestaurants = (offset = 0) => {
	return dispatch => {
		dispatch({
			type: FETCH_RESTAURANTS_LOADING,
			value: true
		});
		return fetchRestaurants(offset)
			.then(restaurants => {
				dispatch({
					type: FETCH_RESTAURANTS_SUCCESS,
					value: restaurants
				});
				dispatch({
					type: QUERY_OFFSET_CHANGED,
					value: offset + OFFSET_INCREMENT
				});
			})
			.catch(err => {
				dispatch(loadError(err));
			});
	};
};

export const getRestaurantDetails = id => {
	return dispatch => {
		dispatch({ type: FETCH_RESTAURANT_DETAILS_LOADING, value: true });
		return fetchRestaurantDetails(id).then(details => {
			dispatch({
				type: FETCH_RESTAURANT_DETAILS_SUCCESS,
				value: details
			});
		});
	};
};

export const getReviews = id => {
	return dispatch => {
		dispatch({ type: FETCH_REVIEWS_LOADING, value: true });
		return fetchReviews(id).then(reviews => {
			dispatch({
				type: FETCH_REVIEWS_SUCCESS,
				value: reviews
			});
		});
	};
};

export const queryOffsetChanged = offset => {
	return {
		type: QUERY_OFFSET_CHANGED,
		value: offset
	};
};

export const openNowChanged = openNow => {
	return {
		type: OPEN_NOW_CHANGED,
		value: openNow
	};
};

export const selectedPriceChanged = selectedPrice => {
	return {
		type: SELECTED_PRICE_CHANGED,
		value: selectedPrice
	};
};

export const selectedCategoryChanged = selectedCategory => {
	return {
		type: SELECTED_CATEGORY_CHANGED,
		value: selectedCategory
	};
};

export const clearRestaurants = () => {
	return {
		type: CLEAR_RESTAURANTS
	};
};

export const clearFilters = () => {
	return {
		type: CLEAR_FILTERS
	};
};
