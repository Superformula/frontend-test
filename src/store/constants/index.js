export const FETCH_RESTAURANTS_LOADING = 'FETCH_RESTAURANTS_LOADING';
export const FETCH_RESTAURANTS_SUCCESS = 'FETCH_RESTAURANTS_SUCCESS';
export const FETCH_REVIEWS_LOADING = 'FETCH_REVIEWS_LOADING';
export const FETCH_REVIEWS_SUCCESS = 'FETCH_REVIEWS_SUCCESS';
export const FETCH_RESTAURANT_DETAILS_LOADING = 'FETCH_RESTAURANT_DETAILS_LOADING';
export const FETCH_RESTAURANT_DETAILS_SUCCESS = 'FETCH_RESTAURANT_DETAILS_SUCCESS';
export const FETCH_ERROR = 'FETCH_ERROR';
export const OPEN_NOW_CHANGED = 'OPEN_NOW_CHANGED';
export const SELECTED_PRICE_CHANGED = 'SELECTED_PRICE_CHANGED';
export const SELECTED_CATEGORY_CHANGED = 'SELECTED_CATEGORY_CHANGED';
export const QUERY_OFFSET_CHANGED = 'QUERY_OFFSET_CHANGED';
export const CLEAR_FILTERS = 'CLEAR_FILTERS';
export const CLEAR_RESTAURANTS = 'CLEAR_RESTAURANTS';
