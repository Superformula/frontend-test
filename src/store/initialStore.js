/**
 * Default store values
 */
export const initialStore = {
  filters: {},
};
