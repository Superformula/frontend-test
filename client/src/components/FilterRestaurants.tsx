import * as React from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { useDispatch, useSelector } from 'react-redux';

import { isOpenToggle, closeOpenToggle, priceItemToggle, catItemToggle, closeFilterLists } from '../redux/actions/actions';
import Dropdown from './Dropdown'
import { DdItem } from '../models/models';

const FilterRestaurants = () => {
  // UseSelector - redux hook that grabs current elements of state within redux store
  const openFilter = useSelector((state: any) => state.openFilter)
  const priceFilter = useSelector((state: any) => state.priceFilter)
  const catFilter = useSelector((state: any) => state.catFilter)
  const ddCats = useSelector((state: any) => state.ddCats)
  const ddPrices = useSelector((state: any) => state.ddPrices)

  const dispatch = useDispatch(); // redux hook that allows actions to be dispatched

  // If default filter setting is selected, set boolean for clear all button to change color
  let enabled: boolean; 
  if (ddCats[0].selected && ddCats[0].content === 'All' && ddPrices[0].selected && ddPrices[0].content === 'All' && openFilter === false) {
    enabled = false;
  } else {
    enabled = true;
  }

  const toggleOpen = () => { // click handler function for toggling open filter
    dispatch(isOpenToggle()) // dispatches isOpenToggle action to redux store
  } 
  
  const clearAll = () => {
    // Pass new array as state with selected property as false to deselect all categories
    const clearCats = ddCats.map((item: DdItem) => {
      if (item.content !== "All") { // Account for All filter edge case
        item.selected = false;
      } else {
        item.selected = true;
      }
      return item;
    });

    // Pass new array as state with selected property as false to deselect all prices
    const clearPrices = ddPrices.map((item: DdItem) => {
      if (item.content !== "All") { // Account for All filter edge case
        item.selected = false;
      } else {
        item.selected = true; // Set All filter to true when the rest are cleared
      }
      return item;
    });
    
    // Dispatch changes to state on redux store
    dispatch(closeOpenToggle()) // Sets Open filter to false
    dispatch(priceItemToggle(clearPrices)) // Pass new array to change state
    dispatch(catItemToggle(clearCats)) // Pass new array to change state
    dispatch(closeFilterLists()) // close current open filter lists
  }

  return (
    <div className="dd-container-wrapper">
      <div className="dd-container">

        <div className="dd-filter">
          {"Filter By:" }
        </div>

        <div className="dd-open" onClick={toggleOpen}>
          {openFilter ? <FontAwesomeIcon icon={['far', 'dot-circle']} color="#002B56" /> : <FontAwesomeIcon icon={['far', 'circle']} color="#002B56" /> }
          {'  Open Now'}
        </div>

        <div className="dd-dd1">
          <Dropdown 
            ddList={ddPrices}
            ddHeader="Price"
            ddOpen={priceFilter}
          ></Dropdown>
        </div>

        <div className="dd-dd2">
          <Dropdown
            ddList={ddCats}
            ddHeader="Categories"
            ddOpen={catFilter}
          ></Dropdown>
        </div>

        <div className="dd-button" >
          {
            enabled ? <button className="dd-btn" onClick={clearAll}> CLEAR ALL </button> :
              <button className="dd-btn" disabled onClick={clearAll}> CLEAR ALL </button>
          }
        </div>

      </div>
    </div>
  )
}

const MemFilterRestaurants = React.memo(FilterRestaurants)

export default MemFilterRestaurants;
