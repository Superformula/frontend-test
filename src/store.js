import { createStore, applyMiddleware } from 'redux';
// using thunk to handle async actions
import thunk from 'redux-thunk';
import produce from 'immer';
import { CHECK_ITEM, CLEAR_FILTERS, UPDATE_RESULTS, getInitialListings } from "./actions";

const initialState = {
  listings: [],
  filters: {},
  query: {},
  limit: 8,
};

const rootReducer = (state = initialState, action) => produce(state, draft => {
    switch (action.type) {
        case CHECK_ITEM:
            draft.filters = action.payload;
            break;
        case CLEAR_FILTERS:
            draft.filters = {};
            break;
        case UPDATE_RESULTS:
            console.log(action.payload);
            draft.listings = action.payload;
            break;
    }
});

const store =  createStore(
    rootReducer,
    initialState,
    applyMiddleware(thunk)
);

getInitialListings(store);

export default store;

