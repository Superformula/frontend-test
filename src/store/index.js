import React, { createContext, useReducer } from 'react';
import { initialStore } from './initialStore';
import { reducer } from './reducer';

/**
 * This is a basic store context implamantation so it can be used to do state control
 * in the website without needing major dependencies such as Redux, Flux, etc.
 */
const StoreContext = createContext();

const StoreProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialStore);

  return (
    <StoreContext.Provider value={[state, dispatch]}>
      {children}
    </StoreContext.Provider>
  );
};

export { StoreContext, StoreProvider };
