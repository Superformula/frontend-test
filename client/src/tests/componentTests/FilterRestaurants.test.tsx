import * as React from 'react'
import * as renderer from 'react-test-renderer';
import { Provider } from 'react-redux'
import { faAngleUp, faAngleDown, faCircle, faCheckCircle } from '@fortawesome/free-solid-svg-icons'
import { faCircle as farCircle, faDotCircle } from '@fortawesome/free-regular-svg-icons'
import { library } from '@fortawesome/fontawesome-svg-core'

library.add(faAngleUp, faAngleDown, faCircle, farCircle, faDotCircle, faCheckCircle)

import FilterRestaurants from '../../components/FilterRestaurants'
import store from '../helpers'

const component: any = renderer.create(
  <Provider store={store}>
    <FilterRestaurants />
  </Provider>
).toJSON()

describe('FilterRestaurants Component', () => {

  // Test initial render
  describe('Prices Dropdown -- Initial Render', () => {

    it('Should render component with all properties without errors', () => {
      expect(component).toMatchSnapshot(); // checks snapshot of loaded JSON data from component
    })

    it('Should render with open now filter not selected', () => { 
      // check if the open filter has an empty circle icon
      const openFilterCheck: boolean = component.children[0].children[1].children[0].props['data-icon'] === 'circle'
      expect(openFilterCheck).toBe(true)
    })

    it('Should render with clear all button disabled', () => {
      // checks button properties to see if it is diabled
      const buttonDisabled: boolean = component.children[0].children[4].children[0].props.disabled
      expect(buttonDisabled).toBe(true)
    })
    
  })
})