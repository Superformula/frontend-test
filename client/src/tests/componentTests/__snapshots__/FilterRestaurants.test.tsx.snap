// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterRestaurants Component Prices Dropdown -- Initial Render Should render component with all properties without errors 1`] = `
<div
  className="dd-container-wrapper"
>
  <div
    className="dd-container"
  >
    <div
      className="dd-filter"
    >
      Filter By:
    </div>
    <div
      className="dd-open"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-circle fa-w-16 "
        color="#002B56"
        data-icon="circle"
        data-prefix="far"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm0 448c-110.5 0-200-89.5-200-200S145.5 56 256 56s200 89.5 200 200-89.5 200-200 200z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
        Open Now
    </div>
    <div
      className="dd-dd1"
    >
      <div
        className="dd-header"
      >
        <div
          className="dd-header-title"
          onClick={[Function]}
        >
          Price  
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-angle-down fa-w-10 "
            color="#C8C8C8"
            data-icon="angle-down"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      className="dd-dd2"
    >
      <div
        className="dd-header"
      >
        <div
          className="dd-header-title"
          onClick={[Function]}
        >
          Categories  
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-angle-down fa-w-10 "
            color="#C8C8C8"
            data-icon="angle-down"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      className="dd-button"
    >
      <button
        className="dd-btn"
        disabled={true}
        onClick={[Function]}
      >
         CLEAR ALL 
      </button>
    </div>
  </div>
</div>
`;
